src/auth/AuthProvider.tsx

```diff
@@ -0,0 +1,49 @@
+import React, {createContext, ReactNode, useCallback, useContext, useState} from 'react'
+
+const AuthContext = createContext<undefined | (AuthContextState & AuthContextActions)>(undefined)
+export const AuthProvider = ({children}: {children: ReactNode}) => {
+  const [state, setState] = useState<AuthContextState>(initialState)
+
+  const actions = {
+    login: useCallback(() => setState(loggedInState), []),
+    logout: useCallback(() => setState(loggedOutState), []),
+  }
+
+  return (
+    <AuthContext.Provider
+      value={{
+        ...state,
+        ...actions,
+      }}
+    >
+      {children}
+    </AuthContext.Provider>
+  )
+}
+
+export const useAuth = () => useContext(AuthContext) || missingProvider()
+
```

Comment:
perhaps?
```suggestion
export const useAuth = () => {
  const auth = useContext(AuthContext)
  if (!auth) missingProvider()
  return {
    isLoggedIn: auth.status === 'logged-in',
    isLoggedOut: auth.status === 'logged-out',
    ...auth
}
```

Reply:
this moves the derivation from the context to the hook, spreading the state into 2 places
there are now 2 types. one for the context, and one for the hook
and it expands the useAuth type to use `boolean` instead of `true` or `false`

Reply:
we would have to rewrite `useAuth` to something like:
```export const useAuth = () => {
  const auth = useContext(AuthContext) || missingProvider()

  return auth.status === 'logged-in'
    ? {
        status: auth.status,
        login: auth.login,
        logout: auth.logout,
        isLoggedIn: true,
        isLoggedOut: false,
      } as const
    : {
        status: auth.status,
        login: auth.login,
        logout: auth.logout,
        isLoggedIn: false,
        isLoggedOut: true,
      } as const
}
```